### PR TITLE
Add two tasks: Listops and IMDB Review

### PR DESCRIPTION
**Pull request by NathanYanJing**
Adding two tasks of LRA benchmark.

### CODE REVIEW

File: s4/s4.py [hunk 1]
```diff
@@ -661,7 +669,7 @@ class StackedModel(nn.Module):
     decode: bool = False
 
     def setup(self):
-        self.encoder = nn.Dense(self.d_model)
+        self.encoder = nn.Embed(num_embeddings=self.n_embed, features=self.d_model)
```
Comment (NathanYanJing) [inline s4/s4.py]: have to change this part of S4 in order to run BN

**Comment by jxiw** [comment]
### ListOps

Downolad dataset in [GitHub page](https://github.com/google-research/long-range-arena). Unzip the downloaded dataset, and move it to the project folder.

```
python -m s4.train --dataset listops-classification --model s4 --epochs 100 --bsz 50 --d_model 128 --n_layers 4 --ssm_n 64 --lr 1e-2 --p_dropout 0 --lr_schedule
```

Training with the previously defined hyper-parameters yields the test accuracy 54.3% on the test set.

### IMDB

We used the huggingface datasets for LRA IMDB review task.

```
python -m s4.train --dataset imdb-classification --model s4 --epochs 100 --bsz 50 --d_model 64 --n_layers 4 --ssm_n 64 --lr 1e-2 --p_dropout 0
```

Training with the previously defined hyper-parameters yields the test accuracy 80.7% on the test set.

**Comment by srush** [comment]
This looks great, but can you make this PR just the new datasets, and not changes to the model or the training?

File: s4/s4.py (hunk 1 above)
Comment (srush) [reply s4/s4.py]: I think this needs to be gated to not break the other code.

**Comment by jxiw** [comment]
Here is the training curve for IMDB Review.

jax-imdb

**Comment by jxiw** [comment]
> 

Ok, working on it.

File: s4/data.py [hunk 2]
```diff
@@ -216,145 +218,6 @@ def create_quickdraw_dataset(bsz=128):
     return trainloader, testloader, N_CLASSES, SEQ_LENGTH, IN_DIM
 
 
-# ### FSDD Sequence Modeling
```
Comment (srush) [inline s4/data.py:219]: I think there was a merge issue. Can you pull the latest main and make sure this part gets back in?

File: s4/s4.py [hunk 3]
```diff
@@ -372,8 +373,8 @@ def K_conv(Ab, Bb, Cb, L):
 
 
 # We can compute the result of applying this filter either with a standard direct convolution or
-# by using convolution theorem with [Fast Fourier Transform (FFT)](https://en.wikipedia.org/wiki/Convolution_theorem). The discrete convolution theorem - for circular convolution of two sequences - allows us to efficiently calculate the output of convolution by first multiplying FFTs of the input sequences and then applying an inverse FFT. To utilize this theorem for non-circular convolutions as in our case, we need to pad the input sequences with zeros, and then unpad the output sequence.
-# As the length gets longer this FFT method will be more efficient than the direct convolution,
+# with a padded (non-circular) [Fast Fourier Transform (FFT)](https://en.wikipedia.org/wiki/Convolution_theorem).
```
Comment (srush) [inline s4/s4.py]: I think there was a merge issue. Makes sure this doesn't get dropped.

File: s4/s4.py [hunk 4]
```diff
@@ -619,18 +620,18 @@ def SSMInit(N):
...
     layer: nn.Module
     l_max: int
     dropout: float
     d_model: int
     training: bool = True
     decode: bool = False
+    pre_norm: bool = True
 
     def setup(self):
+        self.norm = nn.BatchNorm(use_running_average=not self.training)
```
Comment (srush) [inline s4/s4.py]: can you have norm type be an option? Be good not to lose layernorm.

File: s4/s4.py [hunk 5]
```diff
@@ -676,29 +684,19 @@ def setup(self):
...
+        x = self.encoder(x)  # shape batch size x sequence len x hidden dimension
         for layer in self.layers:
             x = layer(x)
         if self.classification:
-            x = np.mean(x, axis=0)
+            x = np.mean(x, axis=1)
         x = self.decoder(x)
-        return nn.log_softmax(x, axis=-1)
-
+        s = nn.log_softmax(x, axis=-1)
```
Comment (srush) [inline s4/s4.py]: keep this as one line.

File: s4/train.py [hunk 6]
```diff
@@ -54,25 +56,54 @@ def map_fn(nested_dict):
     return map_fn
 
 
+# def ReduceLROnPlateau_schedule(
```
Comment (srush) [inline s4/train.py]: Remove

File: s4/train.py [hunk 7]
```diff
@@ -353,25 +386,22 @@ def example_train(
...
+        # )
         if (classification and test_acc > best_acc) or (
-            not classification and test_loss < best_loss
+                not classification and test_loss < best_loss
         ):
             # Create new "best-{step}.ckpt and remove old one
-            shutil.copy(ckpt_path, f"{run_id}/best_{epoch}")
-            if os.path.exists(f"{run_id}/best_{best_epoch}"):
-                os.remove(f"{run_id}/best_{best_epoch}")
+            # shutil.copy(ckpt_path, f"{run_id}/best_{epoch}")
```
Comment (srush) [inline s4/train.py]: put back.

File: s4/train.py [hunk 8]
```diff
@@ -420,10 +450,14 @@ def example_train(
...
 
+    # Max Vocab Length
+    # S4 set max vocab 20 for ListOps, and 135 for IMDB
+    parser.add_argument("--max_vocab", type=int, default=135)
+
     # Weights and Biases Parameters
     parser.add_argument(
         "--use_wandb",
-        default=False,
+        default=True,
```
Comment (srush) [inline s4/train.py]: put back.

**Review by srush — CHANGES_REQUESTED** [review]

**Comment by srush** [comment]
Added some review comments. I think there was a merge failure, so if you can try again to pull. 

Also be sure to run `make autoformat` to check the formatting.

**Comment by NathanYanJing** [comment]
Results of ListOps:

listops

File: s4/s4.py (hunk 1 above)
Comment (NathanYanJing) [reply s4/s4.py]: got it, we will just change the dataset.py for now not to touch the model itself.

File: s4/s4.py (hunk 4 above)
Comment (NathanYanJing) [reply s4/s4.py]: could be hard -- different shapes compared to LN, but can check and try.

Comment (srush) [reply s4/s4.py]: I don't think that's true.
